### ISSUE TITLE
BUG: PIDTEMPBED flaw in Configuration_Store

### ISSUE THREAD

**Issue by RetireeJay46** [prusa3d/Prusa-Firmware#925]
TL;DR - When a user compiles and uploads a version using Bang-Bang control for the heated bed, apparently information is set somewhere that is not under control of the "standard" firmware, and subsequently uploading a version of the firmware that uses PID control causes bed heating to fail completely (no heat comes on at all).  
I somewhat suspect the problem is in ConfigurationStore where apparently PID parameters for the bed are stored or not stored depending on the state of the PIDTEMPBED flag; maybe the presence or absence of these parameters in EEPROM affects the storage and reading of other parameters too since it seems they are stored in a stack with position-dependent meanings.  Perhaps it would be best just to leave the PID parameters in the reading and writing of EEPROM even if they are not used in controlling the bed.

Details on the discovery...
A couple of iterations back, I implemented bang-bang control for my heated bed.  (The 7Hz on/off cycling was creating a "strobe" effect on room lights; the much less frequent cycling of bang-bang control is not noticeable.)
When I installed Version 3.3.0, I got all tangled up with Git, and I had to create a new Repository to get things straightened out (I'm still a total noob with Git).  So I went in to the source code and implemented my mods (by hand) for the changes to ultralcd.cpp for the first-layer filament selection, but I forgot to change the line in Configuration_Prusa.h for heat bed control (//#define PIDTEMPBED).  When I uploaded this firmware to my printer, the bed would not heat!  I subsequently tried "factory" versions 3.3.0,  3.2.1,  3.2.0,  3.1.3,  and 3.1.1 and none of them would heat the bed.  Finally, I corrected the line in Configuration_Prusa.h for version 3.3.0 to use bang-bang control and now my bed heat works.
My electronics is all original, with no modifications at all.

**Comment by RetireeJay46** [comment]
In Configuration_Store, it seems to me that (and I may be displaying ignorance here) if the configuration of data in the EEPROM changes, then you would have a different "version" of the EEPROM.  Changing the Version number would have the desirable effect of forcing the firmware to re-initialize the EEPROM rather than use leftover settings that might be scrambled due to the insertion or deletion of certain "fields" within the EEPROM.  For example, this code for the Hot End maintains the structure of the EEPROM whether or not PID is enabled:

#ifdef PIDTEMP
    EEPROM_WRITE_VAR(i,Kp);
    EEPROM_WRITE_VAR(i,Ki);
    EEPROM_WRITE_VAR(i,Kd);
  #else
		float dummy = 3000.0f;
    EEPROM_WRITE_VAR(i,dummy);
		dummy = 0.0f;
    EEPROM_WRITE_VAR(i,dummy);
    EEPROM_WRITE_VAR(i,dummy);
  #endif`

 Whereas this code for the bed heat does not maintain the structure of the EEPROM data store; it inserts three fields into the stack only if PIDTEMPBED is defined:

 #ifdef PIDTEMPBED
	EEPROM_WRITE_VAR(i, bedKp);
	EEPROM_WRITE_VAR(i, bedKi);
	EEPROM_WRITE_VAR(i, bedKd);
  #endif`

So it seems to me (in my ignorance) that you should either (a) always insert dummy values for unused fields, maintaining the exact structure of the entire EEPROM data set, or (b) re-define the "value" of the EEPROM identification field by calculating it based on the actual sub-fields in use.

**Comment by RetireeJay46** [comment]
I have tested the firmware with #PIDTEMPBED commented out (in Configuration_Prusa.h) and it does seem that the following fixes take care of synchronizing the EEPROM:
#ifdef PIDTEMPBED
	EEPROM_WRITE_VAR(i, bedKp);
	EEPROM_WRITE_VAR(i, bedKi);
	EEPROM_WRITE_VAR(i, bedKd);
  **#else
		float dummy = 3000.0f;
    EEPROM_WRITE_VAR(i,dummy);
		dummy = 0.0f;
    EEPROM_WRITE_VAR(i,dummy);
    EEPROM_WRITE_VAR(i,dummy);**
  #endif

and
#ifdef PIDTEMPBED
	EEPROM_READ_VAR(i, bedKp);
	EEPROM_READ_VAR(i, bedKi);
	EEPROM_READ_VAR(i, bedKd);
**#else
		float dummy;
	EEPROM_READ_VAR(i, dummy);
	EEPROM_READ_VAR(i, dummy);
	EEPROM_READ_VAR(i, dummy);**			
#endif
